### PR TITLE
image: Improve godoc for image types

### PR DESCRIPTION
**Pull request by smarterclayton**
End users should have a bit more info about the core image types.
Describe behavior, intended use, target audience, and some details
of API interactions.

### CODE REVIEW

File: image/v1/types.go [hunk 1]
```diff
@@ -22,6 +22,12 @@ type ImageList struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // Image is an immutable representation of a container image and metadata at a point in time.
+// Images are named by taking a hash of their contents (metadata and content) and any change
+// in format, content, or metadata results in a new name. The images resource is primarily
+// for use by cluster administrators and integrators - end users instead access images via the
+// imagestreamtags or imagestreamimages resources. While image metadata is stored in the API,
+// any integration that accepts images must implement its own storage and use the image
+// resource to manage referential integrity with image streams.
```
Comment (bparees) [inline image/v1/types.go]: i'm not clear on what this last sentence is saying.   s/accepts/manages/ ?  

implements its own storage for what?  the metadata, or for the image content itself?  i would assume the latter.

why does referential integrity to the imagestreams matter?  am i not allowed to just create images(image metadata) without updating the imagestreams?  why not?

File: image/v1/types.go [hunk 2]
```diff
@@ -159,9 +165,20 @@ type ImageStreamList struct {
 // +genclient:method=Layers,verb=get,subresource=layers,result=github.com/openshift/api/image/v1.ImageStreamLayers
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// ImageStream stores a mapping of tags to images, metadata overrides that are applied
+// An ImageStream stores a mapping of tags to images, metadata overrides that are applied
```
Comment (bparees) [inline image/v1/types.go]: have the godoc rules(guidelines?) been loosed to allow articles to preface the type/field/function name?

File: image/v1/types.go [hunk 3]
```diff
@@ -159,9 +165,20 @@ type ImageStreamList struct {
...
+// images which are imported from container registries using credentials in your namespace
+// with the pull secret type, or to existing image stream tags and images which are
+// immediately accessible for tagging or pulling. The history of images applied to a tag
+// is visible in the status.tags field and any user who can view an image stream is allowed
+// to tag that image into their own image streams. Access to pull images from the integrated
+// registry is granted by having the "get imagestreams/layers" permission on a given image
+// stream. Users may remove a tag by deleting the imagestreamtag resource, which causes both
+// spec and status for that tag to be removed. Image stream history is retained until an
+// administrator runs the prune operation, which removes references that are no longer in
+// use. To preserve a historical image, ensure there is a tag in spec pointing to that imag
```
Comment (bparees) [inline image/v1/types.go]:
```suggestion
// use. To preserve a historical image, ensure there is a tag in spec pointing to that image
```

File: image/v1/types.go [hunk 4]
```diff
@@ -331,7 +348,11 @@ type TagEventCondition struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // ImageStreamMapping represents a mapping from a single tag to a container image as
```
Comment (bparees) [inline image/v1/types.go]: single imagestream tag ?

File: image/v1/types.go [hunk 5]
```diff
@@ -331,7 +348,11 @@ type TagEventCondition struct {
...
-// well as the reference to the container image stream the image came from.
+// well as the reference to the container image stream the image came from. This resource
```
Comment (bparees) [inline image/v1/types.go]: how does this mapping map to an imagestream?  it only has an "image" field and a "tag" field?

File: image/v1/types.go [hunk 6]
```diff
@@ -347,6 +368,13 @@ type ImageStreamMapping struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // ImageStreamTag represents an Image that is retrieved by tag name from an ImageStream.
+// Use this resource to interact with the tags and images in an image stream by tag, or
+// to see the image details for a particular tag. The image associated with this resource
+// is the most recently tagged, imported, or pushed image (as described in the image stream
```
Comment (bparees) [inline image/v1/types.go]:
```suggestion
// is the most recent successfully tagged, imported, or pushed image (as described in the image stream
```

and then you can drop the clause about "if an import is in progress or has failed"

File: image/v1/types.go [hunk 7]
```diff
@@ -388,6 +416,14 @@ type ImageStreamTagList struct {
...
 
 // ImageStreamImage represents an Image that is retrieved by image name from an ImageStream.
+// A not found error will be returned if no such image is referenced by a tag within the
+// ImageStream. Images are created when spec tags are set on an image stream that represent
+// an image in an external registry, when pushing to the integrated registry, or when tagging
+// an existing image from one image stream to another. The name of an image stream image is
+// in the form "<STREAM>@<DIGEST>", where the digest is the content addressible identifier
+// for the image (sha256:xxxxx...). You can use ImageStreamImages as the from.kind of an
+// image stream spec tag to reference an image exactly. The only operations supported on
+// the imagestreamimage endpoint are retrieving the image.
```
Comment (bparees) [inline image/v1/types.go]: i think we need more clarification (somewhere) on the distinction between an ImageStreamImage and an Image resource.

File: image/v1/types.go (hunk 2 above)
Comment (smarterclayton) [reply image/v1/types.go]: These are turned into user facing doc, like api type field docs.  So they should be targeted at humans.

File: image/v1/types.go [hunk 8]
```diff
@@ -22,6 +22,12 @@ type ImageList struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // Image is an immutable representation of a container image and metadata at a point in time.
+// Images are named by taking a hash of their contents (metadata and content) and any change
+// in format, content, or metadata results in a new name. The images resource is primarily
+// for use by cluster administrators and integrators - end users instead access images via the
```
Comment (dmage) [inline image/v1/types.go]: Who is an integrator?

File: image/v1/types.go (hunk 2 above)
Comment (bparees) [reply image/v1/types.go]: i mean, i agree, but we also have things like golint that complain about this stuff, right?

Comment (smarterclayton) [reply image/v1/types.go]: Golint should never complain on types.go on that particular error.

File: image/v1/types.go (hunk 6 above)
Comment (smarterclayton) [reply image/v1/types.go]: started a new sentence with the text change

File: image/v1/types.go (hunk 7 above)
Comment (smarterclayton) [reply image/v1/types.go]: Added a second sentence.

File: image/v1/types.go (hunk 8 above)
Comment (smarterclayton) [reply image/v1/types.go]: added more clarifying info.

**Comment by smarterclayton** [comment]
Updated with comments

File: image/v1/types.go [hunk 9]
```diff
@@ -22,6 +22,12 @@ type ImageList struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // Image is an immutable representation of a container image and metadata at a point in time.
+// Images are named by taking a hash of their contents (metadata and content) and any change
+// in format, content, or metadata results in a new name. The images resource is primarily
+// for use by cluster administrators and integrations like the cluster image registry - end
+// users instead access images via the imagestreamtags or imagestreamimages resources. While
+// image metadata is stored in the API, any integration that implements the container image
+// registry API its own storage for the raw manifest data, image config, and layer data.
```
Comment (bparees) [inline image/v1/types.go]:
```suggestion
// registry API must provide its own storage for the raw manifest data, image config, and layer data.
```

**Comment by smarterclayton** [comment]
Fixed, tagging.

/cherry-pick release-4.3

**Pull request merged**